src/NuGet.Core/NuGet.Repositories/NuGetv3LocalRepository.cs

```diff
@@ -2,22 +2,37 @@
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
 using System;
+using System.Collections.Concurrent;
 using System.Collections.Generic;
 using System.IO;
 using System.Linq;
+using NuGet.Packaging;
 using NuGet.Versioning;
 
 namespace NuGet.Repositories
 {
+    /// <summary>
+    /// Caches package info from the global packages folder in memory.
+    /// Packages not yet in the cache will be retrieved from disk.
+    /// </summary>
     public class NuGetv3LocalRepository
     {
-        private readonly Dictionary<string, IEnumerable<LocalPackageInfo>> _cache = new Dictionary<string, IEnumerable<LocalPackageInfo>>(StringComparer.OrdinalIgnoreCase);
+        private readonly ConcurrentDictionary<string, IEnumerable<LocalPackageInfo>> _cache
+            = new ConcurrentDictionary<string, IEnumerable<LocalPackageInfo>>(StringComparer.OrdinalIgnoreCase);
         private readonly bool _checkPackageIdCase;
+        private readonly VersionFolderPathResolver _pathResolver;
+
+        public NuGetv3LocalRepository(string path)
+            : this(path, checkPackageIdCase: false)
+        {
+        }
 
         public NuGetv3LocalRepository(string path, bool checkPackageIdCase)
         {
             RepositoryRoot = path;
             _checkPackageIdCase = checkPackageIdCase;
+
+            _pathResolver = new VersionFolderPathResolver(path);
```

Comment:
If _checkPackageIdCase is false, we don't care about the casing of the package id, right? We could use
VersionFolderPathResolver(path, normalizePackageId: !_checkPackageIdCase).